File: Projects/CPP-VNEXT/src/Animals/Animal.cpp
#include "Animal.h"
#include "../World.h"

Animal::Animal(int strength, int initiative, std::string symbol, Position&& initialPosition) : Organism(strength, initiative,
                                                                                                 symbol,
                                                                                                 std::move(initialPosition)) {

}

void Animal::HandleAction(World& world) {
    std::optional<Position> newPosition = world.GetNearbyPosition(this->GetPosition());
    if (newPosition && world.Move(this->GetPosition(), *newPosition)) {
        this->SetPosition(std::move(*newPosition));
    }
}

void Animal::HandleCollision(CollisionContext& collisionContext) {
    if (collisionContext.GetAttacker()->GetType() == this->GetType()) {
        std::optional<Position> pos = collisionContext.GetWorld().GetNearbyPosition(this->GetPosition(), 1, true);
        if (pos) {
            collisionContext.GetWorld().Log("Breeding", this->GetType() + " has appeared at " + (*pos).ToString());
            Organism* organism = this->GetNewOfType(std::move(*pos));
            collisionContext.GetWorld().Add(organism);
        }
        collisionContext.Cancel();
    }
}

File: Projects/Java/src/main/java/com/nieradko/worldsim/core/animals/Animal.java
package com.nieradko.worldsim.core.animals;

import com.nieradko.worldsim.core.*;
import javafx.scene.paint.Color;

public abstract class Animal extends Organism implements IMovable {

    protected Animal(int strength, int initiative, Color color, Position position) {
        super(strength, initiative, color, position);
    }

    @Override
    protected void handleAction(IActionContext context) {
        context.getRandomNearbyPosition(getPosition())
                .ifPresent(value -> context.move(this, value));
    }

    @Override
    protected void handleCollision(ICollisionContext collisionContext, IWorldContext worldContext) {
        if (collisionContext.getAttacker().getClass() == this.getClass()) {
            var proposedPosition = worldContext.getRandomNearbyPosition(getPosition(), true);
            if (proposedPosition.isPresent()) {
                getNewInstance(proposedPosition.get())
                        .ifPresent(newAnimal -> worldContext.add(newAnimal));
            }
            collisionContext.cancel();
        }
    }

    @Override
    public void moveTo(Position position) {
        setPosition(position);
    }

    public void increaseStrength(int delta) {
        this.strength += delta;
    }
}
